I got this data from maven analytics. It is a mock data and does not contain any missing or duplicate values. 

In [2]:
import pandas as pd

In [3]:
df = pd.read_csv("Bank_Churn.csv")
df

CustomerId    Surname  CreditScore Geography  Gender  Age  Tenure  \
0       15634602   Hargrave          619    France  Female   42       2   
1       15647311       Hill          608     Spain  Female   41       1   
2       15619304       Onio          502    France  Female   42       8   
3       15701354       Boni          699    France  Female   39       1   
4       15737888   Mitchell          850     Spain  Female   43       2   
...          ...        ...          ...       ...     ...  ...     ...   
9995    15606229   Obijiaku          771    France    Male   39       5   
9996    15569892  Johnstone          516    France    Male   35      10   
9997    15584532        Liu          709    France  Female   36       7   
9998    15682355  Sabbatini          772   Germany    Male   42       3   
9999    15628319     Walker          792    France  Female   28       4   

        Balance  NumOfProducts  HasCrCard  IsActiveMember  EstimatedSalary  \
0          0.00              1          1               1        101348.88   
1      83807.86              1          0               1        112542.58   
2     159660.80              3          1               0        113931.57   
3          0.00              2          0               0         93826.63   
4     125510.82              1          1               1         79084.10   
...         ...            ...        ...             ...              ...   
9995       0.00              2          1               0         96270.64   
9996   57369.61              1          1               1        101699.77   
9997       0.00              1          0               1         42085.58   
9998   75075.31              2          1               0         92888.52   
9999  130142.79              1          1               0         38190.78   

      Exited  
0          1  
1          0  
2          1  
3          0  
4          0  
...      ...  
9995       0  
9996       0  
9997       1  
9998       1  
9999       0  

[10000 rows x 13 columns]

In [4]:
df.shape

(10000, 13)

In [5]:
df.sample(5)

CustomerId   Surname  CreditScore Geography  Gender  Age  Tenure  \
6367    15568742    Parkes          536    France  Female   41       9   
9416    15795370    Mazure          648   Germany    Male   37       6   
3627    15609475     Ricci          604     Spain  Female   39       7   
844     15806983      Moss          640    France    Male   44       3   
1920    15723720  McKenzie          591    France  Female   31       7   

        Balance  NumOfProducts  HasCrCard  IsActiveMember  EstimatedSalary  \
6367       0.00              1          1               0        121299.14   
9416  131753.41              1          1               0         86894.67   
3627   98544.11              1          1               1         52327.57   
844   137148.68              1          1               0         92381.01   
1920       0.00              2          0               1         48778.46   

      Exited  
6367       0  
9416       0  
3627       0  
844        0  
1920       0

In [6]:
df.isnull().sum()

CustomerId         0
Surname            0
CreditScore        0
Geography          0
Gender             0
Age                0
Tenure             0
Balance            0
NumOfProducts      0
HasCrCard          0
IsActiveMember     0
EstimatedSalary    0
Exited             0
dtype: int64

In [7]:
df.describe()

CustomerId   CreditScore           Age        Tenure        Balance  \
count  1.000000e+04  10000.000000  10000.000000  10000.000000   10000.000000   
mean   1.569094e+07    650.528800     38.921800      5.012800   76485.889288   
std    7.193619e+04     96.653299     10.487806      2.892174   62397.405202   
min    1.556570e+07    350.000000     18.000000      0.000000       0.000000   
25%    1.562853e+07    584.000000     32.000000      3.000000       0.000000   
50%    1.569074e+07    652.000000     37.000000      5.000000   97198.540000   
75%    1.575323e+07    718.000000     44.000000      7.000000  127644.240000   
max    1.581569e+07    850.000000     92.000000     10.000000  250898.090000   

       NumOfProducts    HasCrCard  IsActiveMember  EstimatedSalary  \
count   10000.000000  10000.00000    10000.000000     10000.000000   
mean        1.530200      0.70550        0.515100    100090.239881   
std         0.581654      0.45584        0.499797     57510.492818   
min         1.000000      0.00000        0.000000        11.580000   
25%         1.000000      0.00000        0.000000     51002.110000   
50%         1.000000      1.00000        1.000000    100193.915000   
75%         2.000000      1.00000        1.000000    149388.247500   
max         4.000000      1.00000        1.000000    199992.480000   

             Exited  
count  10000.000000  
mean       0.203700  
std        0.402769  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max        1.000000

In [8]:
df.duplicated().sum()

0

In [9]:
df_cor = df.select_dtypes(include=['number'])
df_cor.corr()

CustomerId  CreditScore       Age    Tenure   Balance  \
CustomerId         1.000000     0.005308  0.009497 -0.014883 -0.012419   
CreditScore        0.005308     1.000000 -0.003965  0.000842  0.006268   
Age                0.009497    -0.003965  1.000000 -0.009997  0.028308   
Tenure            -0.014883     0.000842 -0.009997  1.000000 -0.012254   
Balance           -0.012419     0.006268  0.028308 -0.012254  1.000000   
NumOfProducts      0.016972     0.012238 -0.030680  0.013444 -0.304180   
HasCrCard         -0.014025    -0.005458 -0.011721  0.022583 -0.014858   
IsActiveMember     0.001665     0.025651  0.085472 -0.028362 -0.010084   
EstimatedSalary    0.015271    -0.001384 -0.007201  0.007784  0.012797   
Exited            -0.006248    -0.027094  0.285323 -0.014001  0.118533   

                 NumOfProducts  HasCrCard  IsActiveMember  EstimatedSalary  \
CustomerId            0.016972  -0.014025        0.001665         0.015271   
CreditScore           0.012238  -0.005458        0.025651        -0.001384   
Age                  -0.030680  -0.011721        0.085472        -0.007201   
Tenure                0.013444   0.022583       -0.028362         0.007784   
Balance              -0.304180  -0.014858       -0.010084         0.012797   
NumOfProducts         1.000000   0.003183        0.009612         0.014204   
HasCrCard             0.003183   1.000000       -0.011866        -0.009933   
IsActiveMember        0.009612  -0.011866        1.000000        -0.011421   
EstimatedSalary       0.014204  -0.009933       -0.011421         1.000000   
Exited               -0.047820  -0.007138       -0.156128         0.012097   

                   Exited  
CustomerId      -0.006248  
CreditScore     -0.027094  
Age              0.285323  
Tenure          -0.014001  
Balance          0.118533  
NumOfProducts   -0.047820  
HasCrCard       -0.007138  
IsActiveMember  -0.156128  
EstimatedSalary  0.012097  
Exited           1.000000

As for EDA, I always used to manually plot everything and try to find the insights. But learning about Pandas Profiling trulty changed it. It is tiem saving and insightful. The below three lines of code provided me a report about my data. I spent hours trying to find insights from the data. It made manual EDA a easier.

In [ ]:
from ydata_profiling import ProfileReport
prof = ProfileReport(df)
prof.to_file(output_file='output.html')

Pandas y-data profiling generated a report of the data. And I found that

1. Most people had fair to good credit scores if I used Vantage Scoring. ( I am thinking of dividing it into categories)

2. Most people are from France and that is about 50% percent of the total given data. Around 25% from Germany and 24.8% from Spain.

3. In the given data 54.6% of people are males and 45.4% are females.

4. People have almost fairly stayed with the company for the tenure of 1 to 9 years but we also have people for less than 1 year and more than 9 years.

5. People from the age group of 30 - 50 have more frequency in the given data. ( I am thinking of dividing it in categories)

6. For some reason people have near to zero balance in their accounts that is about 36.2% of people.

7. Around 50% of people have 1 product of the bank while 45.9% of people have 2 products of the company.

8. About 70.5% have a credit card and 51.5% are active users of the bank.

9. People also have a balance of less than 500 rupees.

10. 20% of people have already churned. These are my insights from the given data. I am thinking of removing the CustomerID and surname column since it is not relevant to my problem statement of building a model and predicting if a customer will churn or not. After these interesting insights from individual columns. I found some correlations as well. Though the correlation is not that strong and lies only in the range of 0 to 0.3 I feel they were better from the correlation of 0.01 and so on.

I will now move on to study these insights better through data visualization. Like I assume people with fewer products associated with bank churn more but I would like to validate it with the data provided. I will also go for feature engineering next, I want to convert credit score and age-group into categories. 